1710: Add SELinux warning controller

jsafrane opened the pull request:
#### What type of PR is this?
/kind feature

### What this PR does / why we need it:
This is a minimal implementation of a controller that watches all Pods on all nodes and warns about two pods using the same volume with different SELinux contexts. Such Pods can't run on the same node - one of them won't be able to access the volume.

The controller does not evaluate any (anti)affinity, it warns even when the Pods can't ever run on the same node.

It implements a simple cache that remember every single volume used by existing Pods and with which SELinux labels it is used. The cache is pruned when a Pod is deleted.

Possible future improvements:
* Don't re-queue Pods whose volumes were already processed. Now, when PV and PVC get bound, both PV and PVC update events cause _all_ Pods in the same namespace to be re-queued and all their volumes re-evaluated.

Changes from the KEP:
* SELinuxController -> SELinuxWarningController. The controller just emits warnings, both as metrics and events.

~WIP:~
* ~It includes https://github.com/kubernetes/kubernetes/pull/127981 to get necessary API, only the last commit is actually this PR.~
* ~Missing unit tests.~
* ~Check SELinuxChangePolicy and warn when two pods use the same volume with different policy. Those won't work on the same node too.~
* ~Add metrics~

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```
Release note in https://github.com/kubernetes/kubernetes/pull/127981 could be enough.

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
https://github.com/kubernetes/enhancements/issues/1710
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Refactor CreateVolumeSpec

Rename old CreateVolumeSpec to CreateVolumeSpecWithNodeMigration that
extracts volume.Spec with node specific CSI migration.

Add CreateVolumeSpec that does the same, only without evaluating node CSI
migration.

jsafrane committed:
Add SELinux warning controller

jsafrane committed:
Fix golint errors

Revealed by the new SELinux warning controller, but not related to it.

gnufied commented:
/assign

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// Add a single volume to the cache. Returns list of conflicts it caused.
+func (c *volumeCache) AddVolume(logger klog.Logger, volumeName v1.UniqueVolumeName, podKey cache.ObjectName, label string, changePolicy v1.PodSELinuxChangePolicy, csiDriver string) []Conflict {
```
can we rename `label` to `selinuxLabel` ?

jsafrane replied:
Renamed.

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+type podInfo struct {
+	// SELinux label to be applied to the volume in the Pod.
+	// Either as mount option or recursively by the container runtime.
+	label string
```
same about renaming `label` to `selinuxLabel`

jsafrane replied:
Renamed

gnufied commented on pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:
```diff
@@ -0,0 +1,507 @@
+	return c, nil
+}
+
+func (c *Controller) addPod(logger klog.Logger, obj interface{}) {
```
should we fix these logger parameters to `_` if we are not using them to prevent linter warnings?

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+}
+
+// GetPodsForCSIDriver returns all pods that use volumes with the given CSI driver.
+func (c *volumeCache) GetPodsForCSIDriver(driverName string) []cache.ObjectName {
```
Isn't this supposed to work for at least couple of non-CSI drivers too?

jsafrane replied:
Evaluation of in-tree volumes are enough at Pod creation time. Only a CSI driver can change its SELinux support in runtime and then we need to re-evaluate all volumes of that driver then.

jsafrane replied:
I added a comment there.

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+// GetPodsForCSIDriver returns all pods that use volumes with the given CSI driver.
+func (c *volumeCache) GetPodsForCSIDriver(driverName string) []cache.ObjectName {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```
Was there any reason to not use `RWMutex` here?

jsafrane replied:
Changed to RWLock, but I think it won't help much. The only read operations are `GetPodsForCSIDriver` (called when CSIDriver `spec.seLinuxMount` changes) and periodic metric scrape `SendConflicts`.

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+	defer c.mutex.Unlock()
+
+	var pods []cache.ObjectName
+	for _, volume := range c.volumes {
```
I wonder if we should have also stored driverName as key somewhere. I hope we don't call this function often.

jsafrane replied:
Maybe as a future optimization. This function is called when CSIDriver `spec.seLinuxMount` changes, to reevaluate all Pods. I think it won't happen that often.

Its complexity is the same as `SendConflicts`, which will be called periodically for each metrics scrape.

gnufied commented on pkg/volume/util/selinux.go:
```diff
@@ -196,3 +234,76 @@ func AddSELinuxMountOption(options []string, seLinuxContext string) []string {
+		// This volume is used with more than one SELinux label in the pod.
+		return info, &MultipleSELinuxLabelsError{labels: labels.UnsortedList()}
+	}
+	if labels.Len() == 0 {
```
If no selinux context is specified for container, then shouldn't the selinux context specified in `PodSecurityContext` take effect?

jsafrane replied:
It does. `GetPodVolumeNames` called by the controller [here](https://github.com/kubernetes/kubernetes/pull/128242/files/34f741e94508605bb31898c48e53e0442a63c8ed#diff-3c16cee6572b0a942684edd7eaf53dcba87418529d3a85bd3fbe42081237feb9R429) calls `DetermineEffectiveSecurityContext` and that already mixes container's SecurityContext and PodSecurityContext together.

gnufied replied:
So it does. Then this function parameter imo is incorrectly named and only reason `podSecurityContext` is supplied here because to determine change policy that needs to be applied. In which case - I would rename `seLinuxContainerContexts` to something like - `effectiveSeLinuxContext` and remove `podSecurityContext` from parameter list and supply `podSecurityContext.SELinuxChangePolicy` directly. But obviously this is not a blocker deal.

jsafrane replied:
Renamed to `effectiveSELinuxContainerLabels`. But I'd prefer to pass `PodSecurityContext`, because I don't need to copy `SELinuxChangePolicy` feature gate evaluation + check for `nil` when dereferencing PodSecurityContext.SELinuxChangePolicy to all callers.

gnufied commented on pkg/controller/volume/selinuxwarning/cache/volumecache.go:
```diff
@@ -0,0 +1,260 @@
+				}
+				if podInfo.label != otherPodInfo.label {
+					ch <- Conflict{
+						PropertyName:       "SELinux label",
```
Since we are reporting this as a metric dimension, I am not sure keeping the whitespace is a good idea.

jsafrane replied:
Changed to `SELinuxLabel` + updated the field description.

jsafrane commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -865,3 +865,43 @@ func startStorageVersionGarbageCollectorController(ctx context.Context, controll
+		name:                names.SELinuxWarningController,
+		aliases:             []string{"selinux-warning"},
+		initFunc:            startSELinuxWarningController,
+		isDisabledByDefault: true,
```
The KEP says:
> This new controller will be opt-in, so it won't run on clusters that don't care about SELinux.

We can change that, still, I think it would be waste of CPU and memory for most Kubernetes distros.

jsafrane committed:
Rework event recorder startup

* Remove Controller.recorder field, there already is eventRecorder.
* Start the event broadcaster in Run(), to save a bit of CPU and memory
  when something initializes the controller, but does not Run() it.
* Log events with log level 3, as the other contollers usually do.
* Use StartStructuredLogging(), which looks fancier than StartLogging

jsafrane commented on pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:
```diff
@@ -0,0 +1,507 @@
+) (*Controller, error) {
+
+	eventBroadcaster := record.NewBroadcaster(record.WithContext(ctx))
+	eventBroadcaster.StartLogging(klog.Infof)
```
I poked around, it starts watching an event broadcaster and whatnot. I moved it to Run, just in case someone instantiates the controller but does not run it, and changed to `StartStructuredLogging`. And changed the log level of events to `V(3)`, as other controllers do.

jsafrane committed:
Rename label to seLinuxLabel

In various parameters, variables and fields. To make the name more
obvious.

jsafrane committed:
Use _ for unused parameters

Sometimes the logger is not used. This fixes some linter warnings.

jsafrane committed:
Use RWLock for the controller cache

It could help a tiny bit with parallel operations.

jsafrane commented:
Rebased to the merged API and fixed most of the review remarks. Working on the rest.

jsafrane commented on pkg/controller/volume/selinuxwarning/selinux_warning_controller.go:
```diff
@@ -0,0 +1,507 @@
+	c.csiTranslator = csiTranslator
+	c.cmpm = csimigration.NewPluginManager(csiTranslator, utilfeature.DefaultFeatureGate)
+
+	err = common.AddPodPVCIndexerIfNotPresent(c.podIndexer)
```
TODO: the indexer needs to move to some common code.

jsafrane replied:
oops, I mis-read the remark. Sure, comment copied.

jsafrane committed:
Add comment how GetPodsForCSIDriver is useful

jsafrane committed:
Add a comment why PVC indexer is used

jsafrane commented:
Fixed gofmt in the first commit.

jsafrane committed:
Update property name in metrics

selinux_volume_conflict should not have space in its label value - it's
harder to query for that value. Use SELinuxLabel as both human friendly (in
an event) and label value.

jsafrane committed:
Describe what the input list of SELinux options is

fedebongio commented:
/remove-sig api-machinery

carlory commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -865,3 +865,43 @@ func startStorageVersionGarbageCollectorController(ctx context.Context, controll
+		return nil, true, fmt.Errorf("failed to probe volume plugins when starting SELinux warning controller: %w", err)
+	}
+
+	ctx = klog.NewContext(ctx, logger)
```
We don't change the logger itself in this fuction, so we can the ctx immediately when creating the controller.

jsafrane replied:
removed

jsafrane committed:
Move feature gate to ControllerDescriptor

jsafrane committed:
Don't create a new controller context

carlory commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/controller_policy.go:
```diff
@@ -501,6 +501,15 @@ func buildControllerRoles() ([]rbacv1.ClusterRole, []rbacv1.ClusterRoleBinding)
+		addControllerRole(&controllerRoles, &controllerRoleBindings, rbacv1.ClusterRole{
+			ObjectMeta: metav1.ObjectMeta{Name: saRolePrefix + "selinux-warning-controller"},
+			Rules: []rbacv1.PolicyRule{
+				eventsRule(),
```
The new controller uses some informers to get pv, pvc, etc. We need to add some permissions for these resources.

jsafrane replied:
I thought that the controller uses shared informers and thus someone else needs those permissions, not the controller itself. But I can see other controllers include them. Added.

jsafrane committed:
Add all RBAC rules for the SELinux controller

The SELinuxWarningController does not necessarily need permissions to read
the objects, because it gets them through a shared informer instantiated by
KCM itself, but let's list the permissions for completeness.

jsafrane commented:
/test pull-kubernetes-e2e-kind
local volume test failed talking to the API server

jsafrane commented on cmd/kube-controller-manager/app/core.go:
```diff
@@ -865,3 +865,43 @@ func startStorageVersionGarbageCollectorController(ctx context.Context, controll
+}
+
+func startSELinuxWarningController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SELinuxChangePolicy) {
```
Good point, node

gnufied commented:
/lgtm
/approve

deads2k commented:
cmd and authorizer changes look ok

/approve

fedebongio commented:
/triage accepted
